src/hb-number-parser.rl

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright © 2019  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ *
+ */
+
+#ifndef HB_NUMBER_PARSER_HH
+#define HB_NUMBER_PARSER_HH
+
+#include "hb.hh"
+
+%%{
+
+machine double_parser;
+alphtype unsigned char;
+write data;
+
+action see_neg { neg = true; }
+action see_exp_neg { exp_neg = true; }
+
+action add_real     { value = value * 10. + (fc - '0'); }
+action add_mantissa { mantissa_factor *= 10; mantissa = mantissa * 10. + (fc - '0'); }
+action add_exp      { exp = exp * 10 + (fc - '0'); }
+
+num = [0-9]+;
+
+main := (
+	(
+		(('+'|'-'@see_neg)? num @add_real) ('.' num @add_mantissa)?
+		|
+		(('+'|'-'@see_neg)? '.' num @add_mantissa)
+	)
+	(('e'|'E') (('+'|'-'@see_exp_neg)? num @add_exp))?
+);
+
+}%%
+
+constexpr double
+pow2 (double x)
+{
+  return x * x;
+}
+
+constexpr double
+pow10_of_2i (unsigned int n)
+{
+  return n == 1 ? 10. : pow2 (pow10_of_2i (n >> 1));
+}
+
+static const double powers_of_10[] =
+{
+  pow10_of_2i (0x100),
+  pow10_of_2i (0x80),
+  pow10_of_2i (0x40),
+  pow10_of_2i (0x20),
+  pow10_of_2i (0x10),
+  pow10_of_2i (0x8),
+  pow10_of_2i (0x4),
+  pow10_of_2i (0x2),
+  pow10_of_2i (0x1),
+};
```

Comment:
Inverting the order here makes more sense to me.

Reply:
The advantage is I can write "for (const double *power = powers_of_10; '''mask'''; ++power, mask >>= 1)"

Reply:
You could still do that, just needed to reverse the order.

Reply:
Oh, ok, let me see